Created main package mujtaba_charm wirh sub packages of analysis, models and utils in it.

MujtabaNasir opened the pull request.

MujtabaNasir committed:
Created main package mujtaba_charm wirh sub packages of analysis, models and utils in it.

mdsrqbl reviewed (changes requested):
- move mujtaba_charm, pyproject, test, readme to root dir
- move analysis, models, utils inside mujtaba_charm
- remove __init__.py from tests dir
- add module docstrings to all init files
- use meaningful branch names
- repository name should be lowercase hyphenated, not snake_case

MujtabaNasir committed:
Moved mujtaba_charm, pyproject, test, readme to root directory.
Moved analysis, models, utils inside mujtaba_charm.
Added docstrings to all init.py files and removed init.py from tests directory.
Repository name changed to mujtaba-charm.

mdsrqbl commented on README.md:
````diff
@@ -1,2 +1,52 @@
+## Installation
+1. To install poetry:
+```
+pip install poetry
````
move this command to some other section
instead write the commands to install this package with pip (after clone and also directly from github)

mdsrqbl commented on pyproject.toml:
```diff
@@ -0,0 +1,14 @@
+readme = "README.md"
+
+[tool.poetry.dependencies]
+python = "^3.12"
```
add support for python 3.8-3.12

mdsrqbl commented on mujtaba_charm/__init__.py:
```diff
@@ -0,0 +1,10 @@
+"""
+This package serves as the main entry point for the mujtaba-charm project. It integrates
```
~package~ module

use proper markdown syntax and add headings and usage examples too

mdsrqbl commented on mujtaba_charm/analysis/__init__.py:
```diff
@@ -0,0 +1,5 @@
+"""
+This package contains modules and functions for data analysis tasks.
```
~package~ sub-package

MujtabaNasir replied:
Isn't mujtaba_charm a main package?

mdsrqbl replied:
this isn't the __init__.py file of the main package

MujtabaNasir replied:
Oh, I miss read. Its in the analysis.

mdsrqbl reviewed (changes requested):
- in the utils subpackage, add a python file with simple hello function (returns string, no print).
- take an optional name parameter in it.
- throw an exception if argument isnt a string
- add proper docstring with usage example
- export the function from the utils sub package
- should work like this

```python
>>> from mujtaba_charm.utils import hello

>>> hello("john")
'hello john!'

>>> hello()
'hello world!'
```

- add a test case for this function (use pytest)
- update readme with package installation and testing commands

MujtabaNasir committed:
I have added the hello function in utils sub-package file named sample.py
Added docstrings
exported the function form the utils sub package
Added 4 test cases using pytest
Updated README.md's usage section, added testing section
Added support for python 3.8-3.12

MujtabaNasir committed:
wrapped all the docstrings lines around 72 characters according to PEP 8 convention.

mdsrqbl commented on README.md:
````diff
@@ -1,2 +1,80 @@
+3. utils
+
+## Installation
+1. Install pytest:
+```bash
+poetry add --dev pytest
+```
+
+2. Ensure all dependencies, including pytest, are installed.
+```bash
+poetry install
+```
````
this is project setup, not package installation. write the command to install `mujtaba_charm`

MujtabaNasir replied:
Fixed

mdsrqbl commented on README.md:
````diff
@@ -1,2 +1,80 @@
+```
+
+## Usage Section
+1. To clone GitHub repo:
````
this section should be about using the package, not about using the tools used to build the package.

MujtabaNasir replied:
1. To clone GitHub repo:
```bash
git clone https://github.com/MujtabaNasir/mujtaba_charm.git
```

2. To install poetry with pip after cloning the repo:
```bash
pip install poetry
```

3. To create and switch to a new branch:
```bash
git checkout -b branch_uno
```

4. To push the new branch to GitHub:
```bash
git push -u origin branch_uno
```

5. To create a poetry package:
```bash
poetry new mujtaba_charm
```

6. Creating analysis, models and utils as sub packages in the main package mujtaba_charm
```bash
mkdir analysis
mkdir models
mkdir utils
```

7. To add __init__.py in sub packages of analysis, models and utils in order to make them python packages
```bash
touch analysis/__init__.py models/__init__.py utils/__init__.py
```

8. To use the hello funtion from the utils sub-package:
```
>>> from mujtaba_charm.utils import hello

>>> hello("john")
'hello john!'

>>> hellp()
'hello world!'
```

So from here,  only point 8 will be in the usage section?

mdsrqbl commented on mujtaba_charm/utils/__init__.py:
```diff
@@ -0,0 +1,7 @@
+data manipulation functions, file handling utilities, etc
+"""
+
+from .sample import hello
```
use absolute imports

MujtabaNasir replied:
Done

mdsrqbl commented on mujtaba_charm/utils/sample.py:
```diff
@@ -0,0 +1,30 @@
+    """
+
+    if not isinstance(name, str):
+        raise TypeError("name should be of string type")
```
also let the user know what data type they passed

MujtabaNasir replied:
Done by using type(name).__name__ to avoid <class int> and just display int

mdsrqbl commented on tests/test_sample.py:
```diff
@@ -0,0 +1,10 @@
+def test_hello():
+    assert hello("john") == "hello john!"
+    assert hello() == "hello world!"
+    assert hello("john") == "hello world!"
```
what is this 3rd test case supposed to check?

MujtabaNasir replied:
It is an invalid case, hello("john") should not == "hello world!", I checked whether this test actually fails like its supposed to.

mdsrqbl commented on tests/test_sample.py:
```diff
@@ -0,0 +1,10 @@
+    assert hello("john") == "hello world!"
+
+    with pytest.raises(TypeError):
+        hello(2024)
```
check if proper keyword was present in the exception message as well.

mdsrqbl reviewed (changes requested):
make the package installable with pip.
make sure it works in editable mode as well.
check the install directory in both of the above cases.

MujtabaNasir committed:
mujtaba_charm is installable with pip and is in editable mode
Fixed usage section in README.md and updated the installation section
Made import path absolute from previous relative path
The user now knows what data type they passed
It checks whether proper keyword is present in the exception message

MujtabaNasir committed:
Deleted .vscode directory

MujtabaNasir committed:
.vscode removed form GitHub
assert hello(123) removed
added .vscode to .gitignore
added pip install git+https://github.com/MujtabaNasir/mujtaba-charm.git under README.md installation

MujtabaNasir committed:
Merge branch 'branch_uno' of https://github.com/MujtabaNasir/mujtaba_charm into branch_uno

MujtabaNasir committed:
Imported utils sub package in main package's __init__.py

mdsrqbl reviewed (approved):
merge the PR
